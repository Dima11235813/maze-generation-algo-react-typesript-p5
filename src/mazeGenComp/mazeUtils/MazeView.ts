import { DEFAULT_Z_DISTANCE } from "../../shared/constants";

export class MazeView {
    constructor(
        public is3d: boolean = true,
        public zValue: number =  0,
        public zoomHeightDiff = DEFAULT_Z_DISTANCE

    ){

    }
}