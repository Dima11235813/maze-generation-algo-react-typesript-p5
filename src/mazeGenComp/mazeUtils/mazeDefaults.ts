import { Color } from "../../utils/colorUtils"

//SET DEFAULT COLORS FOR MAZE
export const mazeDefaultOptions = {
    //SCENE
    defaultBackgroundColor: new Color(145, 101, 100, 100),
    //CELL COLOR
    defaultCellColor: new Color(9, 170, 121, 100),
    defaultCellWallColor: new Color(255, 255, 255, 100),
    //CELL SIZE
    defaultCellSize: 40,
    defaultCellWallSize: 1,
    //CELL WALL SHAPE
    defaultStrokeCapStyle: "ROUND",
}