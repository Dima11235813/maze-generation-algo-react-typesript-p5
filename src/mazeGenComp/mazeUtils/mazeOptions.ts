
import { Color } from "../../utils/colorUtils"
import { mazeDefaultOptions } from "./mazeDefaults";
import { storageUtils } from "../../utils/storageUtils";
import { MazeView } from "./MazeView";
import { MIN_NUMBER_OF_CELLS_HORIZONTALLY, DEFAULT_Z_DISTANCE, ZOOM_MULTIPLIER_DEFAULT } from "../../shared/constants";

export class MazeOptions {
    view: MazeView = new MazeView();
    //DYNAMIC SECTION
    //Window 
    windowHeight: number;
    windowWidth: number;
    calculatedCellHeight: number = 1;
    calculatedCellWidth: number = 1;
    smallerSizeOfCellHeightWidth: number = 1
    aspectRatio: number = 1;
    maxStrokeWidth: number = 1
    padding: number = 1;
    numberOfColumns: number = 10;
    numberOfRows: number = 10;
    constructor(
        //SCENE
        public backgroundColor: Color = mazeDefaultOptions.defaultBackgroundColor,
        //CELL COLOR,
        public cellColor: Color = mazeDefaultOptions.defaultCellColor,
        public cellWallColor : Color= mazeDefaultOptions.defaultCellWallColor,
        //CELL SIZE,
        public cellSize : number = mazeDefaultOptions.defaultCellSize,
        public cellWallSize : number = mazeDefaultOptions.defaultCellWallSize,
        //CELL WALL SHAPE,
        public cellWallStrokeCapStyle: string = mazeDefaultOptions.defaultStrokeCapStyle,
    ) {
        //SET UP DEFAULTS FOR MAZE if not using storage
        //SCENE
        //Update any options that were persisted in storage
        this.updateOptionsFromStorage()
        this.updateDynamicValues()
        //Set UI Vars
        //TODO Move to header constants
        this.windowHeight = window.innerHeight - 90;
        this.windowWidth = window.innerWidth;
        window.addEventListener("wheel", (event: any) => {
            const minZoomSetting = 0
            let userIsZoomingIn = event.wheelDelta > 0
            if (userIsZoomingIn && this.view.zValue < minZoomSetting) {
                this.view.zValue += ZOOM_MULTIPLIER_DEFAULT
            } else {
                this.view.zValue -= ZOOM_MULTIPLIER_DEFAULT
            }
            this.view.zoomHeightDiff = this.windowHeight - this.view.zValue
        })

    }
    saveOptionsToStorage() {
        storageUtils.setMazeoptionsInStorage(this)
    }
    updateOptionsFromStorage() {
        storageUtils.updateMazeOptionsFromStorage(this)
    }
    updateDynamicValues() {
        //DYNAMIC OPTIONS CALC
        if (this.maxStrokeWidth > this.cellSize) {
            this.maxStrokeWidth = this.cellSize
        }
        let ratioFloat = this.windowWidth / this.windowHeight
        this.aspectRatio = parseFloat(ratioFloat.toPrecision(5))
        if (this.aspectRatio > 1.0) {
            this.calculatedCellWidth = this.aspectRatio * this.cellSize
            this.calculatedCellHeight = 1 * this.cellSize
        } else {
            this.calculatedCellWidth = 1 * this.cellSize
            this.calculatedCellHeight = (1 / this.aspectRatio) * this.cellSize
        }
        //calculate stroke width base on either width or height
        this.smallerSizeOfCellHeightWidth = this.calculatedCellHeight > this.calculatedCellWidth ? this.calculatedCellWidth : this.calculatedCellHeight
        this.maxStrokeWidth = this.smallerSizeOfCellHeightWidth / 2;
        //set up number of columns and numberOfRows based on the canvas pixel size and the cell this.windowWidth constants
        this.numberOfColumns = Math.floor(this.windowWidth / this.calculatedCellWidth)
        this.numberOfRows = Math.floor(this.windowHeight / this.calculatedCellHeight)
        // console.log(`MAZE OPTIONS`)
        // console.log(this)
        this.padding = .3 * this.cellSize

        //THIS Was a bad idea 
        // this.padding = Math.floor(this.windowWidth % this.calculatedCellWidth)
    }
}