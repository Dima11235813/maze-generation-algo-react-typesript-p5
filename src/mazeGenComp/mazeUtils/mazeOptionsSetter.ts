import { MazeOptions } from "./mazeOptions"
import { ChangeEvent } from "react"
import { mazeDefaultsStorageKeys, storageUtils } from "../../utils/storageUtils"
import { Color } from "../../utils/colorUtils"
import { logger } from "../../utils/loggingUtils"

export class MazeOptionsSetter {
    constructor(public mazeOptions: MazeOptions) {

    }
    //TODO Move to color utils class
    //SET UP HANDLERS FOR COLOR CHANGE
    updateStorage = () => {
        //todo set up individual setters for options
        storageUtils.setMazeoptionsInStorage(this.mazeOptions)
    }
    handleBackgroundColorChange = (color: any): void => {
        this.assignColorToTarget(color.rgb, this.mazeOptions.backgroundColor)
        this.updateStorage()
    }
    handleCellColorChange = (color: any): void => {
        this.assignColorToTarget(color.rgb, this.mazeOptions.cellColor)
        this.updateStorage()
    }
    handleCellWallColorChange = (color: any): void => {
        this.assignColorToTarget(color.rgb, this.mazeOptions.cellWallColor)
        this.updateStorage()
    }
    assignColorToTarget = (assignmentColor: Color, targetColor: Color) => {
        const { r, g, b, a } = assignmentColor
        targetColor.r = r
        targetColor.g = g
        targetColor.b = b
        //TODO FIGURE OUT WHY SENDING A TO P5 CAUSES ODD THINGS TO HAPPEN
        // if (a) {
        //   this.mazeOptions.cellColor.a = a
        // }
        // if (a) {
        //     targetColor.a = a
        // }
    }
    handleCellWidthChange = (size: number) => {
        this.mazeOptions.cellSize = size
        localStorage.setItem(mazeDefaultsStorageKeys.cellSizeKey, size.toString())
        console.log(`New cell width is ${this.mazeOptions.cellSize}`)
    }


    handleCellWallWidthPercentChange = (event: ChangeEvent<HTMLSelectElement>) => {
        // temp = newWidth
        // newWidth += temp
        // if (newWidth > minCellWidth && newWidth < maxCellWidth) {
        //   mazeOptions.cellSize = newWidth
        // } else {
        // }
        let newWidthPercent = parseInt(event.target.value)
        this.mazeOptions.cellWallSize = newWidthPercent
        logger(`New cell wall width percent is ${this.mazeOptions.cellWallSize}`)
    }
}