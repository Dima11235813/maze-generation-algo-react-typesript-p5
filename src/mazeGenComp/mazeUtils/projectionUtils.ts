import { mazeOptionsUiContext } from "../../AppContext"
import { MazeOptions } from "./mazeOptions"
import { useContext } from "react"
import { Point } from "../components/Point"


export class ProjectionTo3dUtilClass {
    constructor() {

    }
}
export const offsetWidthBy3dProjection = (value: number, mazeOptions: MazeOptions): number => {
    const { windowWidth,
    } = mazeOptions
    let offsetWidth3d = Math.floor(value - (windowWidth / 2))
    // console.log(offsetWidth3d)
    return offsetWidth3d
}
export const offsetHeightBy3dProjection = (value: number, mazeOptions: MazeOptions): number => {
    const {
        windowHeight,
    } = mazeOptions
    let offsetHeight3d = Math.floor(value - (windowHeight / 2))
    // console.log(offsetHeight3d)
    return offsetHeight3d
}



export const getProjectionFor3D = (use3d: boolean, point: Point, mazeOptions: MazeOptions) => {
    let x_value = point.x
    let y_value = point.y
    //set up point vals based on 2D or 3D projection 
    //TODO extract this logic
    let xColPointValToDraw
    let yRowPointValToDraw
    if (use3d) {
        xColPointValToDraw = offsetWidthBy3dProjection(x_value, mazeOptions)
    } else {
        xColPointValToDraw = x_value
    }
    if (use3d) {
        yRowPointValToDraw = offsetHeightBy3dProjection(y_value, mazeOptions)
    } else {
        yRowPointValToDraw = y_value
    }
    let pointToReturn = new Point(xColPointValToDraw, yRowPointValToDraw)
    console.log(pointToReturn)
    return pointToReturn
}