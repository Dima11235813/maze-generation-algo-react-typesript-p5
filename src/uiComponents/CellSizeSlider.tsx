import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Slider from "@material-ui/core/Slider";
import Input from "@material-ui/core/Input";
import FormatSize from "@material-ui/icons/FormatSize";

//Maze Config
import { MazeOptions } from "../mazeGenComp/mazeUtils/mazeOptions";
import { MazeOptionsSetter } from "../mazeGenComp/mazeUtils/mazeOptionsSetter";
import { Labels } from "../shared/labels";

interface CellSizeSlider {
  mazeOptionsSetter: MazeOptionsSetter;
  onSizeChange: Function;
  cellSize: number;
  windowWidth: number

}

const CellSizeSlider = (props: CellSizeSlider) => {
  const { mazeOptionsSetter, windowWidth, cellSize, onSizeChange } = props!;
  //cell width rules
  let minCellSize = 6;
  let maxCellSize = windowWidth / 20;
  let cellSizeIncrementor = 2;

  //create ui component
  const useStyles = makeStyles({
    root: {
      width: 200
    },
    input: {
      width: 42
    }
  });
  const classes = useStyles();

  //Set up state for slider
  const [value, setValue] = React.useState(cellSize);

  //Event for slider
  const handleSliderChange = (event: any, newValue: any) => {
    setStateAndPersistSelection(newValue);
  };
  //Even for input
  const handleInputChange = (event: any) => {
    let newValue =
      event.target.value === "" ? minCellSize : Number(event.target.value);
    setStateAndPersistSelection(newValue);
  };
  const setStateAndPersistSelection = (newValue: number) => {
    let valueForStorage = setStateConditionally(newValue);
    updateCellSizeForMaze(valueForStorage);
  };
  const updateCellSizeForMaze = (newValue: number) => {
    mazeOptionsSetter.handleCellSizeChange(newValue);
    onSizeChange();
  };

  //state change

  const handleBlur = () => {
    setStateConditionally(value);
  };
  const setStateConditionally = (value: number): number => {
    if (value < minCellSize) {
      setValue(minCellSize);
      return minCellSize;
    } else if (value > maxCellSize) {
      setValue(maxCellSize);
      return maxCellSize;
    } else {
      setValue(value);
      return value;
    }
  };

  return (
    <div className={classes.root}>
      <Grid container spacing={2} alignItems="center">
        <Grid item>
          <FormatSize />
        </Grid>
        <Grid item xs>
          <Slider
            value={typeof value === "number" ? value : minCellSize}
            onChange={handleSliderChange}
            aria-labelledby="input-slider"
          />
        </Grid>
        <Grid item>
          <Input
            className={classes.input}
            value={value}
            margin="dense"
            onChange={handleInputChange}
            onBlur={handleBlur}
            inputProps={{
              step: cellSizeIncrementor,
              min: minCellSize,
              max: maxCellSize,
              type: "number",
              "aria-labelledby": "input-slider"
            }}
          />
        </Grid>
      </Grid>
    </div>
  );
};
export default CellSizeSlider;
