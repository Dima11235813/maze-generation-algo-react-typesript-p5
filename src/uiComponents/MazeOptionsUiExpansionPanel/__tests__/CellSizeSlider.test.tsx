import React from 'react';
import CellSizeSlider, { CellSizeSliderProps } from '../../CellSizeSlider';
import { MazeOptions } from '../../../mazeGenComp/mazeUtils/mazeOptions';
import { MazeOptionsSetter } from '../../../mazeGenComp/mazeUtils/mazeOptionsSetter';
import { shallow } from 'enzyme';

let mazeOptions = new MazeOptions()

const cellSizeSliderProps: CellSizeSliderProps = {
    mazeOptionsSetter: new MazeOptionsSetter(mazeOptions),
    onSizeChange: () => undefined,
    cellSize: 5,
    windowWidth: 500,

}

it('renders without crashing', () => {
    let wrapper = shallow(<CellSizeSlider {...cellSizeSliderProps} />);
    expect(wrapper).toMatchSnapshot()
});
