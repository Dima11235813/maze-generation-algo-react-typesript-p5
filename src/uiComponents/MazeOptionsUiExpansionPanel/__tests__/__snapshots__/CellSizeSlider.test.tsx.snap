// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <FormatSizeIcon />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={true}
    >
      <WithStyles(ForwardRef(Slider))
        aria-labelledby="input-slider"
        onChange={[Function]}
        value={5}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Input))
        className="makeStyles-input-2"
        inputProps={
          Object {
            "aria-labelledby": "input-slider",
            "max": 55.55555555555556,
            "min": 3,
            "step": 1,
            "type": "number",
          }
        }
        margin="dense"
        onBlur={[Function]}
        onChange={[Function]}
        value={5}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
